<a href="https://colab.research.google.com/github/scitech-butterfly/Restaurant_Ratings/blob/main/Restaurant_Rating_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Zomato Restaurant Data Analysis**

Importing the dataset and performing basic analytical operations

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from google.colab import files

In [ ]:
uploaded = files.upload()

Saving zomato.csv to zomato.csv


In [ ]:
import io

df = pd.read_csv(io.BytesIO(uploaded['zomato.csv']))

In [ ]:
df.head()

Unnamed: 0.1  Unnamed: 0     restaurant name restaurant type  \
0             0           0        #FeelTheROLL     Quick Bites   
1             1           1          #L-81 Cafe     Quick Bites   
2             2           2             #refuel            Cafe   
3             3           3  '@ Biryani Central   Casual Dining   
4             4           4          '@ The Bbq   Casual Dining   

   rate (out of 5)  num of ratings  avg cost (two people) online_order  \
0              3.4               7                  200.0           No   
1              3.9              48                  400.0          Yes   
2              3.7              37                  400.0          Yes   
3              2.7             135                  550.0          Yes   
4              2.8              40                  700.0          Yes   

  table booking                                      cuisines type  \
0            No                                          Fast Food   
1            No                               Fast Food, Beverages   
2            No                                    Cafe, Beverages   
3            No                          Biryani, Mughlai, Chinese   
4            No  BBQ, Continental, North Indian, Chinese, Bever...   

                             area      local address  
0                       Bellandur          Bellandur  
1  Byresandra,Tavarekere,Madiwala                HSR  
2               Bannerghatta Road  Bannerghatta Road  
3                    Marathahalli       Marathahalli  
4                       Bellandur          Bellandur

In [ ]:
df.describe()

Unnamed: 0.1   Unnamed: 0  rate (out of 5)  num of ratings  \
count   7105.000000  7105.000000      7037.000000     7105.000000   
mean    3552.000000  3552.000000         3.514253      188.921042   
std     2051.181164  2051.181164         0.463249      592.171049   
min        0.000000     0.000000         1.800000        1.000000   
25%     1776.000000  1776.000000         3.200000       16.000000   
50%     3552.000000  3552.000000         3.500000       40.000000   
75%     5328.000000  5328.000000         3.800000      128.000000   
max     7104.000000  7104.000000         4.900000    16345.000000   

       avg cost (two people)  
count            7048.000000  
mean              540.286464  
std               462.902305  
min                40.000000  
25%               300.000000  
50%               400.000000  
75%               600.000000  
max              6000.000000

In [ ]:
df.isnull().sum()

Unnamed: 0.1              0
Unnamed: 0                0
restaurant name           0
restaurant type           0
rate (out of 5)          68
num of ratings            0
avg cost (two people)    57
online_order              0
table booking             0
cuisines type             0
area                      0
local address             0
dtype: int64

In [ ]:
df['rate (out of 5)'].fillna(df['rate (out of 5)'].median(), inplace=True)

# Replace null values in 'average_cost_for_two' with the mean of 'average_cost_for_two'
df['avg cost (two people)'].fillna(df['avg cost (two people)'].median(), inplace=True)

In [ ]:
df.head()

Unnamed: 0.1  Unnamed: 0     restaurant name restaurant type  \
0             0           0        #FeelTheROLL     Quick Bites   
1             1           1          #L-81 Cafe     Quick Bites   
2             2           2             #refuel            Cafe   
3             3           3  '@ Biryani Central   Casual Dining   
4             4           4          '@ The Bbq   Casual Dining   

   rate (out of 5)  num of ratings  avg cost (two people) online_order  \
0              3.4               7                  200.0           No   
1              3.9              48                  400.0          Yes   
2              3.7              37                  400.0          Yes   
3              2.7             135                  550.0          Yes   
4              2.8              40                  700.0          Yes   

  table booking                                      cuisines type  \
0            No                                          Fast Food   
1            No                               Fast Food, Beverages   
2            No                                    Cafe, Beverages   
3            No                          Biryani, Mughlai, Chinese   
4            No  BBQ, Continental, North Indian, Chinese, Bever...   

                             area      local address  
0                       Bellandur          Bellandur  
1  Byresandra,Tavarekere,Madiwala                HSR  
2               Bannerghatta Road  Bannerghatta Road  
3                    Marathahalli       Marathahalli  
4                       Bellandur          Bellandur

In [ ]:
df.isna().sum()

Unnamed: 0.1             0
Unnamed: 0               0
restaurant name          0
restaurant type          0
rate (out of 5)          0
num of ratings           0
avg cost (two people)    0
online_order             0
table booking            0
cuisines type            0
area                     0
local address            0
dtype: int64

# **Recommendation System:**
## On the basis of cuisine type and average cost for two

In [ ]:
pr_cuisine = input("Enter your preference of cuisine: ")
pr_rating = float(input("Enter the average cost rating for two: "))  # Assuming the rating is a float value

# Filter the DataFrame to include only rows that match the user's criteria
filtered_df = df[(df['cuisines type'].str.contains(pr_cuisine, case=False)) & (df['avg cost (two people)'] == pr_rating)]

if not filtered_df.empty:
    # Sort the filtered DataFrame by ratings in descending order
    sorted_df = filtered_df.sort_values(by='rate (out of 5)', ascending=False)

    # Print the top 15 restaurants with the highest ratings
    top_restaurants = sorted_df.head(15)
    for index, row in top_restaurants.iterrows():
        print(f"{row['restaurant name']} - Rating: {row['rate (out of 5)']}")
else:
    print("No restaurants found matching the criteria.")


Enter your preference of cuisine: north indian
Enter the average cost rating for two: 400
Mumbai Tiffin - Rating: 4.4
Aligarh House - Rating: 4.4
Boozers Bar & Kitchen - Rating: 4.3
Kota Kachori - Rating: 4.3
Raat Ki Rani - Rating: 4.3
nu.tree - Rating: 4.3
Chandni Chowk 2 Bangalore - Rating: 4.2
The Paratha Company - Rating: 4.2
Mooch Marod - Rating: 4.2
New Agarwal Bhavan - Rating: 4.1
The Good Grain - Rating: 4.1
Hotel Salam - Rating: 4.1
Udupi Vaibhav - Rating: 4.1
Khazana Food Paradise - Rating: 4.1
Lazeez Biryani - Rating: 4.0


# **Machine Learning Model**
Training the model to predict the restaurant ratings based on 5 different columns and finding out the features with the highest impact on ratings

In [ ]:
import sklearn as sk
from sklearn.ensemble import RandomForestRegressor
from pandas import read_csv
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_absolute_error

selected_cols = ['avg cost (two people)', 'online_order', 'table booking', 'cuisines type', 'area', 'rate (out of 5)']
df_selected = df[selected_cols]
x = df_selected[['avg cost (two people)', 'online_order', 'table booking', 'cuisines type', 'area' ]]
y =  df.iloc[0:, 4:5].values
column_trans = ColumnTransformer(
    [('encoder', OneHotEncoder(), ['online_order', 'table booking', 'cuisines type', 'area' ])],
    remainder='passthrough')

x_encoded = column_trans.fit_transform(x)
print(x.shape, y.shape)
x_train, x_test, y_train, y_test = train_test_split(x_encoded, y, test_size=0.33, random_state=1)
y_train = y_train.ravel()
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)
sk.ensemble.RandomForestRegressor
model = RandomForestRegressor(random_state=1)
model.fit(x_train, y_train)
yhat = model.predict(x_test)
print(yhat)
mae = mean_absolute_error(y_test, yhat)
print('MAE: %.3f' % mae)

(7105, 5) (7105, 1)
(4760, 2210) (2345, 2210) (4760,) (2345, 1)
[3.691 3.612 3.479 ... 3.431 3.207 3.528]
MAE: 0.339


In [ ]:
# Get feature importances from the trained model
importances = model.feature_importances_

# Get the names of the features
feature_names = column_trans.get_feature_names_out()

# Create a DataFrame to store feature importances
feature_importances_df = pd.DataFrame(data={'Feature': feature_names, 'Importance': importances})

# Sort feature importances in descending order
feature_importances_df = feature_importances_df.sort_values(by='Importance', ascending=False)

# Print the correlation between each feature and ratings
feature_importances_df.head(20)


Feature  Importance
2209                   remainder__avg cost (two people)    0.142803
2                             encoder__table booking_No    0.106156
3                            encoder__table booking_Yes    0.037300
0                              encoder__online_order_No    0.017403
1459                encoder__cuisines type_North Indian    0.013436
2185       encoder__area_Byresandra,Tavarekere,Madiwala    0.011978
1520       encoder__cuisines type_North Indian, Chinese    0.011367
2180                    encoder__area_Bannerghatta Road    0.010900
2184                          encoder__area_Brookefield    0.009811
2187                      encoder__area_Electronic City    0.009696
1981                encoder__cuisines type_South Indian    0.009359
2182                            encoder__area_Bellandur    0.009326
2201                         encoder__area_Malleshwaram    0.009160
2190                          encoder__area_Indiranagar    0.009153
2179                         encoder__area_Banashankari    0.008424
2202                         encoder__area_Marathahalli    0.008275
2183                         encoder__area_Brigade Road    0.007482
2189                                  encoder__area_HSR    0.007044
2054  encoder__cuisines type_South Indian, North Ind...    0.006051
2193                         encoder__area_Kalyan Nagar    0.005925

# **Hypothesis Testing**
## **Z-Test**

Null Hypothesis: The mean rating of restaurants with average cost (for two) less than or equal to 400 is equal to the mean rating of restaurants with average cost for two above 400.


Alternative Hypothesis: The mean rating of restaurants with average cost for two less than or equal to 400 is greater than the mean rating of restaurants with average cost for two above 400.

Level of Significance: 95% or 0.95

Critical Value: 1.96


In [ ]:
def calculate_mean_std(df):
    mean_rate = df['rate (out of 5)'].mean()
    std_rate = df['rate (out of 5)'].std()
    return mean_rate, std_rate


In [ ]:
def calculate_mean_std_condition(df, condition):
    if condition == 'less_than_400':
        filtered_data = df[df['avg cost (two people)'] <= 400]
    elif condition == 'greater_than_400':
        filtered_data = df[df['avg cost (two people)'] > 400]
    else:
        print("Invalid condition")
        return None

    mean_rate, std_rate = calculate_mean_std(filtered_data)
    return mean_rate, std_rate


In [ ]:
# Calculate mean and standard deviation for average cost less than or equal to 400
mean_less_than_400, std_less_than_400 = calculate_mean_std_condition(df, 'less_than_400')
print("Mean rate for average cost less than or equal to 400:", mean_less_than_400)
print("Standard deviation of rate for average cost less than or equal to 400:", std_less_than_400)

# Calculate mean and standard deviation for average cost greater than 400
mean_greater_than_400, std_greater_than_400 = calculate_mean_std_condition(df, 'greater_than_400')
print("\nMean rate for average cost greater than 800:", mean_greater_than_400)
print("Standard deviation of rate for average cost greater than 400:", std_greater_than_400)


Mean rate for average cost less than or equal to 400: 3.413301362948006
Standard deviation of rate for average cost less than or equal to 400: 0.398085870200995

Mean rate for average cost greater than 800: 3.6412026726057904
Standard deviation of rate for average cost greater than 400: 0.5018090297340028


In [ ]:
def calculate_mean_std_condition(df, condition):
    if condition == 'less_than_400':
        filtered_data = df[df['avg cost (two people)'] <= 400]
    elif condition == 'greater_than_400':
        filtered_data = df[df['avg cost (two people)'] > 400]
    else:
        print("Invalid condition")
        return None

    mean_rate, std_rate = calculate_mean_std(filtered_data)
    count = filtered_data.shape[0]  # Count of rows
    return mean_rate, std_rate, count


In [ ]:
# Calculate mean and standard deviation for average cost less than or equal to 400
mean_less_than_400, std_less_than_400, count_less_than_400 = calculate_mean_std_condition(df, 'less_than_400')
print("Mean rate for average cost less than or equal to 400:", mean_less_than_400)
print("Standard deviation of rate for average cost less than or equal to 400:", std_less_than_400)
print("Count of average cost less than or equal to 400:", count_less_than_400)

# Calculate mean and standard deviation for average cost greater than 400
mean_greater_than_400, std_greater_than_400, count_greater_than_400 = calculate_mean_std_condition(df, 'greater_than_400')
print("\nMean rate for average cost greater than 400:", mean_greater_than_400)
print("Standard deviation of rate for average cost greater than 400:", std_greater_than_400)
print("Count of average cost greater than 400:", count_greater_than_400)


Mean rate for average cost less than or equal to 400: 3.413301362948006
Standard deviation of rate for average cost less than or equal to 400: 0.398085870200995
Count of average cost less than or equal to 400: 3962

Mean rate for average cost greater than 400: 3.6412026726057904
Standard deviation of rate for average cost greater than 400: 0.5018090297340028
Count of average cost greater than 400: 3143


In [ ]:
import math
x=mean_less_than_400-mean_greater_than_400
s=(((std_less_than_400)**2)/count_less_than_400) + ((std_greater_than_400**2)/count_greater_than_400)
y=math.sqrt(s)

In [ ]:
t=x/y
print("Test Statistic: ", t)

Test Statistic:  -20.794354289538216


The absolute value of the test statistic is 20.79 which is greater than the critical value. Hence, it lies in the region of rejection.